docs/cluster-configuration.md

```diff
@@ -299,9 +305,13 @@ It allows the platform to compromise performance/availability of VMIs to guarant
 
 ### allowPostCopy
 
-It enables post-copy live migrations. Such migrations allow even the busiest VMIs to successfully live-migrate.
-However, events like a network failure can cause a VMI crash.
-If set to true, migrations will still start in pre-copy, but switch to post-copy when CompletionTimeoutPerGiB triggers.
+When enabled, KubeVirt attempts to use post-copy live-migration in case it 
+reaches its completion timeout while attempting pre-copy live-migration.
+Post-copy migrations allow even the busiest VMs to successfully live-migrate.
+However, events like a network failure or a failure in any of the source or
+destination nodes can cause the migrated VM to crash or reach inconsistency.
```

Comment:
/hold

Until I figure out whether the VM can actually reach an inconsistent state in case the destination node crashes.

Reply:
A network failure should not crash either side of the VM, but put both sides into a so-called "postcopy-paused" state.  QEMU allows to recover from such state back to "postcopy-active" with the help of an administrator after v3.0.  The problem is that if the current network is down or broken, the admin may need to provide an optional channel to move on the migration until complete.  For a rough idea on how it works (including QEMU APIs) please refer to this page:

https://wiki.qemu.org/Features/PostcopyRecovery

I'm not familiar enough with Libvirt, but I think the corresponding API for libvirt is "--postcopy-resume", as described in the Libvirt documentation:

https://libvirt.org/manpages/virsh.html

Reply:
I'm not sure what's the best policy here for kubevirt, but a dumb one can be that when postcopy hit network issue (then both QEMUs will go into "postcopy-paused" state), we set a timeout, trying to reconnect using the same old URI to destination QEMU after timeout reached, and keep doing so if it keeps failing until succeed again.  I assume then there's no admin involvement needed if the network can be recovered already automatically via the same route (I'm mostly considering something like an accidentally unplugged wire).

Besides, the admin can overwrite the URI anytime for the case where the old route will be completely broken.  It can be some separate interface besides above auto-reconnect timeout policy, maybe.

Reply:
@xzpeter thanks. This complicates things quite a bit. @vladikr , how does KubeVirt behave in this case?

@xzpeter can you elaborate on what happens to the source or destination if its counterpart dies? To me, the only consistent action is to crash.

Reply:
@xzpeter yes, I agree. A crash is a reasonable and acceptable result. My problem is with other modes of failure, e.g

> If src QEMU died, since some pages still reside on the src QEMU then dest QEMU will have no way to complete such migration. Any access to missing pages will be permanently blocked hence same to a VM crash.

Do you mean that dst would stay paused forever? Who is responsible to kill it so that a fresh VM can restart (maybe KubeVirt already does it)?

Reply:
> Do you mean that dst would stay paused forever? Who is responsible to kill it so that a fresh VM can restart (maybe KubeVirt already does it)?

Yes it will.  That's the best QEMU can do, IMHO.  It will also be the same case no matter src or dest QEMU died. 

When any side of QEMU crashed, the other side of QEMU should logically go into "postcopy-paused" state just like when there is a network issue - consider when one QEMU died and process recycled, the socket resources will be released by the kernel of that host, so it can have a similar effect of having the socket disconnected to the peer.

Here no matter which side got down, the management layer may need to make the decision on killing the other side of QEMU.  Mgmt apps can have such knowledge of "other side QEMU crashed" by e.g. capturing child exit signals, but QEMU doesn't have that knowledge.  That's why QEMU stays in PAUSED state is what it does right now, as the best effort to maintain whatever data still resides within the QEMU, and with the hope that it could be reconnected in the near future so that data won't get lost.

Reply:
/hold cancel

We need to check this better in KubeVirt. I filed https://issues.redhat.com/browse/CNV-38195 to track it, and added a warning here. I think this PR can be merged now.